dash/dash.py

```diff
@@ -662,11 +662,13 @@ def _validate_callback(self, output, inputs, state, events):
                             name.lower(), str(arg), name
                         ))
 
-                if '.' in arg.component_id:
+                invalid_characters = ['.', ':', '[', ']']
+                if any(x in arg.component_id for x in invalid_characters):
                     raise exceptions.IDsCantContainPeriods('''The element
-                    `{}` contains a period in its ID.
+                    `{}` contains {} in its ID.
                     Periods are not allowed in IDs right now.'''.format(
```

Comment:
It says `Periods` regardless of the invalid character, also this will now break any apps that used `[, ], :` as an ID character for a single output. I think that we would instead need to check if `.` is in the ID for all callbacks, then check if `[`. `]`, `:` is in the ID for all callbacks where `isinstance(output, (list, tuple))`.

Reply:
If we're going to make any breaking change we might as well give ourselves freedom for future expandability - thinking about #475 in particular but who knows what else may come up.

What if we required IDs to be valid Python identifiers? ie letters, numbers, digits, and underscore, first character not a digit. That would enable other possibilities like collecting all components into a namespace somewhere - eg `app.components_by_id.myDiv3`. Dunno what that would be useful for, but the point is "valid Python identifier" is a broad enough class that people shouldn't feel constricted by it, and it's instantly recognizable to Python devs so folks won't need to hunt around for the rule as to which characters are allowed or not.